### ISSUE TITLE
Development mode flag

### ISSUE THREAD

**Issue by djarek** [otland/forgottenserver#2215]
We need to have a development mode flag (defualted to true) in ```config.lua``` that enables/disables unsafe features and issues warnings in some situations, e.g.
- server reloads should be disabled if not in development mode
- binding to IPs other than ```127.0.0.1``` or ```::1``` in development mode should issue a **BIG SCARY WARNING**

**Comment by diath** [comment]
>server reloads should be disabled if not in development mode

I disagree, the reload system is a great way to deploy a hotfix without resorting to restarting the server when not necessary.

**Comment by djarek** [comment]
> I disagree, the reload system is a great way to deploy a hotfix without resorting to restarting the server when not necessary.

The problem is that it can silently corrupt the game state if something goes wrong.

**Comment by ranisalt** [comment]
Will do!

> binding to IPs other than 127.0.0.1 or ::1 in development mode should issue a BIG SCARY WARNING

I'm in favor of **completely** disabling unsafe features on production mode, as for the user to modify source code and be completely up to him/her if anything goes wrong. It makes very explicit that the source code has been altered and thus we the devs are not responsible for any issues.

**Comment by DSpeichert** [comment]
I disagree. Forcing people to modify source code to get things that worked before to work again is going to introduce more problems. It won't make it explicit that source code has been altered. They won't disclose that on support board. Devs are already "not responsible" for any issues. Software should not be crippled to try to save people who will immediately remove that "protection". User education has its place in software documentation (that we are lacking somewhat), not source code.

**Comment by ranisalt** [comment]
At least make it annoying, then: if config is unsafe, oblige users to type "yes", else halt.

**Comment by DSpeichert** [comment]
That breaks all kinds of automation, hence is like breaking the code. I get that you're trying to force a certain behavior but from what I found out, that's a moot point. People will remove these safety checks and break more in the process. And complain, they always complain anyway.

**Comment by ranisalt** [comment]
You shouldn't automate a broken, unsafe procedure. E.g. if we have a docker script for the server, it should boot the server on a clean, safe and properly configured manner.

Anyway, I'm open to that. We can strict requirements in the future, if wanted.

**Comment by WibbenZ** [comment]
I get that we wanna make TFS as stable and as easy to run as posible, but things like this should be easy to use.
Let's say you wrote a perfect script (regarding the coding) but you ex wrote getHEalth insted of getHealth with Player:getHealth() should we really need to inform the players about an upcoming (unannounced) restart due to a small error like that?

IMO it should be set to what kind of coder you are, not if you are in dev mode or not.
Sure we can all do mistakes, but we should stll be able to patch them quickly without restarts that takes 5-15min (with broadcasts etc)

And TBF I think it's good as it is right now, people will learn that reloads can "ruin" your server and avoid it, when you advance you will learn when you can use it.

This will just make it harder for devs making them sticking to an older version just like 0.x.

**Comment by ranisalt** [comment]
I don't think we are talking about typos on scripts. It's more failure to load a file or unsafe database configurations.

**Comment by WibbenZ** [comment]
Yes but if the reload function is blocked in "safe mode" we can't change simple things like that without restarting the server.
So IMO the reload function should have some scary messages sure, but it shouldn't be disabled or be harder to  use then today.

**Comment by ranisalt** [comment]
Ideally on production your server won't need urgent changes, right? If you are hotfixing scripts, you are developing.

**Comment by WibbenZ** [comment]
Well there are things that you might need to fix quickly, ex if someone found a typo in an NPC should you restart, wait till the next daily SS or do a reload and be done with it in a second?

It's things like this that will annoy ppl and make them run with dev mode at all times.
But I agree that we need more documentation about it, or maybe alot of scary messages when you use the reload functions, but ye it shouldn't be disabled.

**Comment by ranisalt** [comment]
Maybe section reloads between safe and unsafe ones? Reloading NPC configuration like name and outfit isn't a problem, the only problem is touching Lua script system.

**Comment by WibbenZ** [comment]
What I ment with a typo was the dialog you have with the NPC = Lua.
Not sure what a good way to "solve" this would be ...

In a way I think we should remove it because people complain about crashes etc and havn't learned anything from it.
But then I also think it will make it harder to do quick fixes and things like that.

If I hosted a server and needed to restart to do a hot fix I would but the server in dev mode at the next restart to avoid it, but I kinda also know what is safe to reload and if I should risk it or not.

**Comment by ranisalt** [comment]
It's impossible to make it safe. As you know, some issues will result in corrupted memory access and crash the server instantly. You have to put on the balance if you prefer to crash the server due to a corruption or gently restart it.

**Comment by WibbenZ** [comment]
Yes to make it safe, but to know if I can reload X without the server crashing.
Sure if I had 1k players I wouldn't reload anything but very few have that amount of players.

I really don't know what a good way to "solve" this is, at the same time if you make it a config boolean people will either by default set it to true or maybe forget to put it to false when they update the server or start hosting it.

**Comment by ranisalt** [comment]
I don't think we have to go fuzzy here. Either it is safe or it is not, there's no "it's probably safe if X and Y".

**Comment by WibbenZ** [comment]
Thats the problem, there are things that are safe to reload as you said npc xmls, confih.lua etc

Atleast in my experience

**Comment by Zbizu** [comment]
> > I disagree, the reload system is a great way to deploy a hotfix without resorting to restarting the server when not necessary.
> 
> The problem is that it can silently corrupt the game state if something goes wrong.

That's the responsibility server owners and scripters should take because only thing that bugs out is lua scripts in this case. All it takes to secure important data from reloading is just typing `if not variable then varible = something end`.

To deploy a hotfix to variable protected this way they can just comment the `if not` statement and reload.
